### PR TITLE
Convert types to sealed classes where possible

### PR DESCRIPTION
**Pull request by dain**
Convert DecimalType, TimeWithTimeZoneType, and TimestampWithTimeZoneType to sealed classes

## Release notes

(X) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/type/LongTimeWithTimeZoneType.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
 import static java.lang.String.format;
 import static java.lang.invoke.MethodHandles.lookup;
 
-class LongTimeWithTimeZoneType
+non-sealed class LongTimeWithTimeZoneType
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/type/LongTimeWithTimeZoneType.java:39]: Why non-sealed? These should be final, instead.

Comment (dain) [reply core/trino-spi/src/main/java/io/trino/spi/type/LongTimeWithTimeZoneType.java:39]: I have no idea... IntelliJ auto converted :(

**Review by martint — APPROVED** [review]

**Comment by hashhar** [comment]
What's the motivation? 

Does this prevent a plugin from providing specialized types for example?

**Comment by findepi** [comment]
> Does this prevent a plugin from providing specialized types for example?

These classes were not meant to be extended and it would be dangerous if someone tried that.

File: core/trino-spi/src/main/java/io/trino/spi/type/TimestampWithTimeZoneType.java [hunk 2]
```diff
@@ -22,9 +22,10 @@
  * @see ShortTimestampWithTimeZoneType
  * @see LongTimestampWithTimeZoneType
  */
-public abstract class TimestampWithTimeZoneType
+public abstract sealed class TimestampWithTimeZoneType
         extends AbstractType
         implements FixedWidthType
+        permits LongTimestampWithTimeZoneType, ShortTimestampWithTimeZoneType
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/type/TimestampWithTimeZoneType.java:28]: The `@see LongTimestampWithTimeZoneType` above became redundant

**Pull request merged**